gems/smithy-client/lib/smithy-client/param_converter.rb

```diff
@@ -0,0 +1,252 @@
+      @converters = Hash.new { |h, k| h[k] = {} }
+
+      def initialize(rules)
+        @rules = rules
+        @opened_files = []
+      end
+
+      attr_reader :opened_files
+
+      # @param [Hash] params
+      # @return [Hash]
+      def convert(params)
+        structure(@rules, params)
+      end
+
+      def close_opened_files
+        @opened_files.each(&:close)
+        @opened_files = []
+      end
+
+      private
+
+      def structure(shape, values)
+        values = c(shape, values)
+        if values.is_a?(::Struct) || values.is_a?(Hash)
+          values.each_pair do |k, v|
+            next if v.nil?
+            next unless shape.member?(k)
+
+            values[k] = member(shape.member(k), v)
+          end
+        end
+        values
+      end
+
+      def union(shape, values)
+        values = c(shape, values)
+        if values.is_a?(Union) || values.is_a?(Hash)
+          values.each_pair do |k, v|
+            next if v.nil?
+            next unless shape.member?(k)
+
+            values[k] = member(shape.member(k), v)
+          end
+        end
+        values
+      end
+
+      def list(shape, values)
+        values = c(shape, values)
+        if values.is_a?(Array)
+          values.map { |v| member(shape.member, v) }
+        else
+          values
+        end
+      end
+
+      def map(shape, values)
+        values = c(shape, values)
+        if values.is_a?(Hash)
+          values.each.with_object({}) do |(key, value), hash|
+            hash[member(shape.key, key)] = member(shape.value, value)
+          end
+        else
+          values
+        end
+      end
+
+      def member(member_shape, value)
+        shape = member_shape.shape
+        case shape
+        when StructureShape then structure(shape, value)
+        when UnionShape then union(shape, value)
+        when ListShape then list(shape, value)
+        when MapShape then map(shape, value)
+        else c(shape, value)
+        end
+      end
+
+      def c(ref, value)
+        self.class.c(ref.class, value, self)
+      end
+
+      class << self
+        def convert(shape, params)
+          new(shape).convert(params)
+        end
+
+        # Registers a new value converter. Converters run in the context
+        # of a shape and value class.
+        #
+        #     # add a converter that stringifies integers
+        #     shape_class = Shapes::StringShape
+        #     ParamConverter.add(shape_class, Integer) { |i| i.to_s }
+        #
+        # @param [Class<Shapes::Shape>] shape_class
+        # @param [Class] value_class
+        # @param [#call] block An object that responds to `#call`
+        #    accepting a single argument. This function should perform
+        #    the value conversion if possible, returning the result.
+        #    If the conversion is not possible, the original value should
+        #    be returned.
+        # @return [void]
+        def add(shape_class, value_class, &block)
```

Comment:
Should this be part of the public interface?

I guess I could see a user wanting to use custom data types and registering them this way, but in general it seems a little bit of an anti-pattern to support through global state.